### ISSUE TITLE
When viewing a data log currently being recorded, have a way to manually refresh to bring in the latest points.

### ISSUE THREAD

**Issue by pdille** [CMU-CREATE-Lab/flutter-app-android#200]
Also add an indicator that it is still being recorded. (same spinner from the main listing page?)

I'd also say that when recording has finished perhaps force a refresh and then hide the spinner/button.

(Adding you Steve, but feel free to punt back if you don't want this one)

**Comment by SteveFulton95** [comment]
Resolved in 42d1461d6dc5311f5a140094c5a805652ff2e2c5.

To refresh the list, I used a swipe refresh layout.  So you just pull down on the list and it will refresh it.  This may not be obvious for the students, so if this needs changed just re-open and let me know.

Also, when I force the log to be updated, I use the SaveToKindle Dialog so the user is also forced to save the data log to the kindle.  If we do not want this functionality either, let me know.

**Comment by pdille** [comment]
Played with it just now:

- First time I clicked on a log from the listing page that was currently mid record I got the loading popup and it stayed there, never going away even after recording finished. Couldn't reproduce since...
- Second time I tried it didn't get stuck and brought up the log. 
- At this point I can swipe down to have it refresh. Two problems: 
1. we don't have this behavior anywhere else so no one would know to do this. At the very least add text at the top to tell the user to do this.
2. it's also not clear that it's actually still recording. Perhaps put a small spinner and text: "Swipe down to pull in most recent data" (or something like that) at the top?

As for when it finishes recording to force the user to save to the Kindle, I'm okay with that. Only happens if you have it open at the time, which is good.

Thoughts on this, @ehamner ?

**Comment by SteveFulton95** [comment]
That spinner holding up the app was there before this feature. Not sure when it started. I noticed it a few days ago and completely forgot about it. I meant to add a separate issue.

I'll  fix the issues asap.

On Jun 7, 2017 2:46 PM, pdille <notifications@github.com> wrote:

Played with it just now:

  *   First time I clicked on a log from the listing page that was currently mid record I got the loading popup and it stayed there, never going away even after recording finished. Couldn't reproduce since...
  *   Second time I tried it didn't get stuck and brought up the log.
  *   At this point I can swipe down to have it refresh. Two problems:

  1.  we don't have this behavior anywhere else so no one would know to do this. At the very least add text at the top to tell the user to do this.
  2.  it's also not clear that it's actually still recording. Perhaps put a small spinner and text: "Swipe down to pull in most recent data" (or something like that) at the top?

As for when it finishes recording to force the user to save to the Kindle, I'm okay with that. Only happens if you have it open at the time, which is good.

Thoughts on this, @ehamner<https://github.com/ehamner> ?

**Comment by SteveFulton95** [comment]
I put a spinner and text telling the user to pull down to pull the most recent data like @pdille suggested in commit 21c57eb81f0429290dd36148490febd34afe61da.  If we decide we want to do it differently, reopen it once again.

**Comment by pdille** [comment]
I like this. Some more things I noticed.:

1. If I start a new recording and look at an old log, it tells me to swipe to get more data. We should only see this on a log that's currently being recorded.

2. Is it possible to disable the action of swiping down when the log is done recording? The message and spinner does go away but I can still swipe down and see the android refresh icon.

3. The values and the icon headers are not aligned. I think they shifted when the view changed to the swipe one or maybe I'm just going crazy and they always looked like this. But anyway, would be great to make the values centered aligned there.

**Comment by SteveFulton95** [comment]
Resolved in commits 0460f814c54bb992d9ae1dd073503d67e3d47dea, 5bedb2e56a4fa4a1f3a931f4ad8a235f7ee84b3d and fc56384b3ed2b9ac0e97fd969f6b9be439e2cdd2.  Third time's a charm. haha

**Comment by pdille** [comment]
^ Awesome. :)

Something else though. When I go back and forth between the data log listing and a data log, it keeps telling me (through the toast message) that a log finished recording. But no recording is happening.

**Comment by SteveFulton95** [comment]
That is happening since I refactored how I was handling states.  I noticed this but thought it was fine because it displays that toast when the log that has been recorded has not been saved.  So I thought it would be okay as kind of a reminder that you can now save it to the kindle. So should I still handle that case anyway?

**Comment by SteveFulton95** [comment]
To clarify, it only displays the toast when you navigate to where you can open a log.

**Comment by pdille** [comment]
Another thing I noticed is that when I saved a log it opened it but the android refresh icon is stuck spinner forever.
![screenshot_2017-06-14-12-19-34](https://user-images.githubusercontent.com/933363/27143303-ec474bfa-50fb-11e7-9e6c-43418468ef0d.png)

**Comment by pdille** [comment]
In regards to the toast message:

1. After playing around a bit more I've grown to like the reappearing toast message. As you mentioned, it's a gentle reminder.

2. That said, there is some oddness with the 'Open Log' button. The way it is now, everytime I click it and then click away it shows that message. If it showed the message when the pop-up appears that would be good (like what happens when I go to the 'main' log list), but the message appears only after loading a new set. To me, that message should be shown again only as a reminder when viewing the full list and not after a log has been opened. Let me know if this wasn't clear. :)

3. I swear the toast appeared before when on different screens (Robot/Sensors). It doesn't seem to do that anymore and it would be great to get that back.

**Comment by SteveFulton95** [comment]
So I won't close this yet, haha. In commits 4b529e948028ea65cebc12855302bcd595a597b8 and 99a4f1d658a1a3cbf537d715255815d1031692ba I fixed the refresh icon (I think) and the toast being shown after you select a data log from the open dialog.

Personally, I was not able to replicate the refresh icon being stuck there.  However, where I add/remove the enabling of the swipe refresh layout, I also make sure to set if it is refreshing to true/false respectively.

In regards to the toast appearing on the Robot/Sensors tab, I was unable to replicate it.  However, it could be because of some synchronization of methods that handle the different states I made.  For example, I have noticed that the toast sometimes displays a couple of seconds after the spinner has already been removed.  So if you navigated away from the data logs tab right around the time the log would be finished recording, this could potentially happen.  I haven't tested this extensively, though.  If we continue to see this, I could probably fix it through the onPause method of the data logs tab.

**Comment by pdille** [comment]
Haha, sorry for being so nitpicky here.

In regards to commit '99a4f1d' to deal with the toast and selecting logs. Is it possible to show the toast when the "Open Log" modal comes up AND there's a log not yet saved? This would be similar to how it behaves when I click on the Data Log tab and initially get to the listing. Basically, there's two ways to see the listing and I'd like this toast to be consistent across them.

As for the toast appearing across tabs, I actually want that behavior. If I start recording and go to the Robot tab for awhile, it would be cool to see a reminder that my log is done. That way I don't have to keep going back to the Data Log tab to check on things. I swear this worked before but perhaps I'm misremembering.

**Comment by SteveFulton95** [comment]
@pdille  Do we want the open log dialog to behave just like the main listing?  Currently, the dialog "pauses" the streaming of data points.  So when you click Open Log, it takes a snapshot of the current data logs and will not update the data log that is currently being recorded.  Do we want the dialog to update data logs the same way?

**Comment by SteveFulton95** [comment]
@pdille @tasota So that toast used to be appear across tabs, but that caused a sensor streaming issue #204. To fix this bug, I made sure pause updating the data logs when OnPause has been called.

In the Session class we have an instance of FlutterMessageListener which determines how exactly we react based off of the responses received from the Flutter.  With this design, it prohibits us from being able to have multiple listeners waiting for responses (sensor streaming and updating data points).  So unless I completely override the pattern that we have been using, I don't think it is possible to achieve this behavior without some sort of backlash.  Unless I am misunderstanding the structure of how Session works. Thoughts?

**Comment by tasota** [comment]
Thanks, Steve. I will send an email shortly and we can discuss this in more detail. I think this can become its own issue

**Comment by pdille** [comment]
@SteveFulton95: I'm fine with how the pop-up works for Open Log. No need for it to live update and show a spinner there. The only change I was hoping for was that when you do click it and the pop-up comes up it displays a toast saying that your Flutter is done recording. Of course only do this if there is a log sitting on the Flutter. This is just like the behavior you'd see when you click on the tab (in regards to the toast). Does that make sense?

**Comment by SteveFulton95** [comment]
The toast message now appears cross different activities and when you click open log since commit c001a02cb25a0db9f9eca9f6ce42228c648d852c.

**Comment by SteveFulton95** [comment]
I am going to close this since all of the points mentioned by @pdille have been resolved.  Feel free to reopen the issue! :)
